src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -108,25 +142,89 @@ private async Task<IEnumerable<RazorCodeAction>> GetCodeActionsAsync(CodeActionP
+
+            return filteredCSharpCodeActions;
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> FilterCSharpCodeActionsAsync(RazorCodeActionContext context, IEnumerable<RazorCodeAction> csharpCodeActions, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+               return null;
+            }
+
+            var tasks = new List<Task<RazorCodeAction[]>>();
+
+            foreach (var provider in _csharpCodeActionProviders)
+            {
+                var result = provider.ProvideAsync(context, csharpCodeActions, cancellationToken);
+                if (result != null)
+                {
+                    tasks.Add(result);
+                }
+            }
+
+            return await ConsolidateCodeActionsFromProvidersAsync(tasks, cancellationToken);
+        }
+
+        private async Task<IEnumerable<RazorCodeAction>> GetCSharpCodeActionsFromLanguageServerAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            Range projectedRange = null;
+            if (context.Request.Range != null &&
+                !_documentMappingService.TryMapToProjectedDocumentRange(
+                    context.CodeDocument,
+                    context.Request.Range,
+                    out projectedRange))
+            {
+                return Array.Empty<RazorCodeAction>();
+            }
+
+            context.Request.Range = projectedRange;
+
             if (cancellationToken.IsCancellationRequested)
             {
                return null;
             }
 
-            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
-            var linePosition = new LinePosition((int)request.Range.Start.Line, (int)request.Range.Start.Character);
-            var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
-            var location = new SourceLocation(hostDocumentIndex, (int)request.Range.Start.Line, (int)request.Range.Start.Character);
+            var response = _languageServer.SendRequest(LanguageServerConstants.RazorProvideCodeActionsEndpoint, context.Request);
+            return await response.Returning<RazorCodeAction[]>(cancellationToken);
+        }
 
-            var context = new RazorCodeActionContext(request, documentSnapshot, codeDocument, location, _languageServerFeatureOptions.SupportsFileManipulation);
-            var tasks = new List<Task<RazorCodeAction[]>>();
+        private async Task<IEnumerable<RazorCodeAction>> GetRazorCodeActionsAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            var linePosition = new LinePosition(
+                context.Request.Range.Start.Line,
+                context.Request.Range.Start.Character);
+            var hostDocumentIndex = context.SourceText.Lines.GetPosition(linePosition);
+            context.Location = new SourceLocation(
```

Comment:
Why not do this when the context is constructed? Having the location be settable seems odd only for this.

Reply:
There was a reason. Replying to this comment, I realize that reasoning was invalid. Fixing 😄 

We only needed location for `RazorCodeAction` so I didn't want to compute it for both, but considering we generate the context only once for both, there wouldn't be any recomputation.